src/bin/pg_autoctl/cli_do_monitor.c

```diff
@@ -161,26 +165,173 @@ keeper_cli_monitor_get_primary_node(int argc, char **argv)
+		{ "help", no_argument, NULL, 'h' },
+		{ NULL, 0, NULL, 0 }
+	};
+
+	optind = 0;
+
+	/* see comments in cli_common.c in function keeper_cli_getopt_pgdata() */
+	unsetenv("POSIXLY_CORRECT");
+
+	while ((c = getopt_long(argc, argv, "D:JVvqh",
+							long_options, &option_index)) != -1)
+	{
+		switch (c)
+		{
+			case 'D':
+			{
+				strlcpy(options.pgSetup.pgdata, optarg, MAXPGPATH);
+				log_trace("--pgdata %s", options.pgSetup.pgdata);
+				break;
+			}
+
+			case 'J':
+			{
+				outputJSON = true;
+				log_trace("--json");
+				break;
+			}
+
+			case 'V':
+			{
+				/* keeper_cli_print_version prints version and exits. */
+				keeper_cli_print_version(argc, argv);
+				break;
+			}
+
+			case 'v':
+			{
+				++verboseCount;
+				switch (verboseCount)
+				{
+					case 1:
+						log_set_level(LOG_INFO);
+						break;
+
+					case 2:
+						log_set_level(LOG_DEBUG);
+						break;
+
+					default:
+						log_set_level(LOG_TRACE);
+						break;
+				}
+				break;
+			}
+
+			case 'q':
+			{
+				log_set_level(LOG_ERROR);
+				break;
+			}
+
+			case 'h':
+			{
+				commandline_help(stderr);
+				exit(EXIT_CODE_QUIT);
+				break;
+			}
+
+			default:
+			{
+				/* getopt_long already wrote an error message */
+				errors++;
+			}
+		}
+	}
+
+	if (errors > 0)
+	{
+		commandline_help(stderr);
+		exit(EXIT_CODE_BAD_ARGS);
+	}
+
+	if (IS_EMPTY_STRING_BUFFER(options.pgSetup.pgdata))
+	{
+		char *pgdata = getenv("PGDATA");
+
+		if (pgdata == NULL)
+		{
+			log_fatal("Failed to get PGDATA either from the environment "
+					  "or from --pgdata");
+			exit(EXIT_CODE_BAD_ARGS);
+		}
+
+		strlcpy(options.pgSetup.pgdata, pgdata, MAXPGPATH);
+	}
+
+	log_debug("Managing PostgreSQL installation at \"%s\"",
+			  options.pgSetup.pgdata);
+
+	if (!keeper_config_set_pathnames_from_pgdata(&options.pathnames,
+												 options.pgSetup.pgdata))
+	{
+		/* errors have already been logged */
+		exit(EXIT_CODE_BAD_ARGS);
+	}
+
+	if (!file_exists(options.pathnames.config))
+	{
+		log_fatal("Expected configuration file does not exists: \"%s\"",
+				  options.pathnames.config);
+
+		if (!directory_exists(options.pgSetup.pgdata))
+		{
+			log_warn("HINT: Check your PGDATA setting: \"%s\"",
+					 options.pgSetup.pgdata);
+		}
+
+		exit(EXIT_CODE_BAD_ARGS);
+	}
+
+	keeperOptions = options;
+
+	return optind;
 }
 
 
 /*
- * keeper_cli_monitor_get_other_node contacts the pg_auto_failover monitor and
+ * keeper_cli_monitor_get_other_nodes contacts the pg_auto_failover monitor and
  * retrieves the "other node" information for given nodename and port.
+ *
+ * TODO: add a --json output, an array of NodeAddress objects.
  */
 static void
-keeper_cli_monitor_get_other_node(int argc, char **argv)
+keeper_cli_monitor_get_other_nodes(int argc, char **argv)
 {
 	KeeperConfig config = keeperOptions;
 	bool missing_pgdata_is_ok = true;
 	bool pg_is_not_running_is_ok = true;
 
 	Monitor monitor = { 0 };
-	NodeAddress otherNode = { 0 };
+
+	/* arbitrary limit to 12 other nodes */
```

Comment:
hardcoded limit gives me itches. 12 is pretty generous upper limit, but still can we do without it ?

Reply:
I pondered the alternative, but that would require to introduce careful memory management, which we avoided up to now. I would much rather than an artificial limit or 12 nodes (for now) rather than deal with dynamic memory in many places...